codec/unknownproto/unknown_fields.go

```diff
@@ -89,37 +104,46 @@ func (ckr *Checker) RejectUnknownFields(b []byte, msg proto.Message) error {
 				// TYPE_BYTES and TYPE_STRING as per
 				// https://github.com/gogo/protobuf/blob/5628607bb4c51c3157aacc3a50f0ab707582b805/protoc-gen-gogo/descriptor/descriptor.go#L95-L118
 			default:
-				return fmt.Errorf("failed to get typename for message of type %v, can only be TYPE_STRING or TYPE_BYTES", typ)
+				return hasUnknownNonCriticals, fmt.Errorf("failed to get typename for message of type %v, can only be TYPE_STRING or TYPE_BYTES", typ)
 			}
 			continue
 		}
 
 		// Let's recursively traverse and typecheck the field.
 
+		// consume length prefix of nested message
+		_, o := protowire.ConsumeVarint(fieldBytes)
+		fieldBytes = fieldBytes[o:]
+
 		if protoMessageName == ".google.protobuf.Any" {
 			// Firstly typecheck types.Any to ensure nothing snuck in.
-			if err := ckr.RejectUnknownFields(fieldBytes, (*types.Any)(nil)); err != nil {
-				return err
+			hasUnknownNonCriticalsChild, err := RejectUnknownFields(fieldBytes, &types.Any{}, allowUnknownNonCriticals)
```

Comment:
Why the change to &types.Any{}? We just need the proto.Message definition to retrieve descriptors, and this incurs an unnecessary allocation.

Reply:
Okay that makes sense